packages/expo/cli/start/server/UrlCreator.ts

```diff
@@ -0,0 +1,159 @@
+import assert from 'assert';
+import { URL } from 'url';
+
+import * as Log from '../../log';
+import { getIpAddress } from '../../utils/ip';
+
+export interface CreateURLOptions {
+  /** URL scheme to use when opening apps in custom runtimes. */
+  scheme?: string | null;
+  /** Type of dev server host to use. */
+  hostType?: 'localhost' | 'lan' | 'tunnel';
+  /** Requested hostname. */
+  hostname?: string;
+}
+
+export class UrlCreator {
```

Comment:
The `new URL` calls in this class will throw if URLs are malformed. Do we want this (fail fast) or do you think it is better to warn and return null as if the URL weren't specified?

Reply:
It's mixed, in some cases we want a null so we can fallback on a less optimal solution like opening from a bundle id.